{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div style=\"text-align: center;\">\n",
    "<h1>CS 614 Algorithms in Bioinformatics - Fall 2020</h1>\n",
    "<h1 >Assignment #3 </h1>\n",
    "</div>\n",
    "\n",
    "**Homework Information:** Some of the problems are probably too long to be done the night before the due date, so plan accordingly. Late assignments will be penalized by a 10% per day. No homework will be accepted more than one week late. Feel free to ask for questions and help, but **the work you submit in should be your own.** It is very important that you enter your answers in the provided cells. You can add extra cells to explore approaches, but only the provided cell can and will be graded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace the following string values with the requested information\n",
    "# Leave the quotation marks surronding the string!\n",
    "class Student:\n",
    "    first = \"firstName\"\n",
    "    last = \"lastName\"\n",
    "    sid = \"StudentID\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Problem 1 Motif profile (10 points) </h3>\n",
    "<p>\n",
    "The following 10 sequences contain a motif of length 8:   \n",
    "<p style=\"font-family: Courier,monospace;\">\n",
    "GAACTCATGGTG<br>\n",
    "AAAAGCACGGTC<br>\n",
    "TCAAAGCAAGGC<br>\n",
    "CCTAATCAGGGC<br>\n",
    "AAGTATGGACTC<br>\n",
    "ACTAAGCAGGGT<br>\n",
    "TCTCACGGCCCA<br>\n",
    "CCTCGTGGTGGG<br>\n",
    "TACCGTATGGTT<br>\n",
    "ACCACTCGTCGA<br>\n",
    "</p>\n",
    "<p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5> Problem 1.1 </h5>\n",
    "<p>\n",
    "Find the highest-score motif of length 8 in the 10 sequences shown above. Construct an alignment of all instances of the motif and compute the profile and consensus sequence of the motif (In the event of a “tie”, select one nucleotide as a representative for that position).\n",
    "    \n",
    "<p>\n",
    "Hint: you may use the functions in the example ipynb's in the class. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This cell is provided for testing your code. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5>Enter your answers here. THIS CELL WILL BE GRADED. </h5>\n",
    "\n",
    "What is the alignment of all instances of the motif?\n",
    "\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>\n",
    "\n",
    "What is the profile of this motif?\n",
    "\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>\n",
    "\n",
    "What is the consensus of this motif?\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>\n",
    "\n",
    "What is the entropy score of the motif profile?\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5>Problem 1.2</h5>\n",
    "Compute the likelihood ratio of getting “ACTCATGG” according to motif profile vs. a background model of each base having equal probability of 25%. (Show the steps or code)\n",
    "<p>\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Problem 2 Sequence Assembly (15 points)</h3>\n",
    "In this assignment, you will try out sequence assembly using Eulerian path algorithms respectively. You will need a set of 15 base-pair simulated reads from a genome that will be used for assembly. The cell below provides various functions and imports necessary for this problem set. Do not import any addtional packages. Also, make sure that you run the following cell. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import math\n",
    "%matplotlib inline\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def loadFasta(fp):\n",
    "    \"\"\" Parses a classically formatted FASTA file into two lists. \n",
    "        One of headers and a second list of sequences.\n",
    "        The ith index of each list correspond.\"\"\"\n",
    "    # split at headers\n",
    "    data = fp.read().decode('utf-8').split('>')\n",
    "    # ignore whatever appears before the 1st header\n",
    "    data.pop(0)     \n",
    "    headers = []\n",
    "    sequences = []\n",
    "    for sequence in data:\n",
    "        lines = sequence.split('\\n')\n",
    "        headers.append(lines.pop(0))\n",
    "        sequences.append(''.join(lines))\n",
    "    return (headers, sequences)\n",
    "\n",
    "class Graph:\n",
    "    def __init__(self, vlist=[]):\n",
    "        \"\"\" Initialize a Graph with an optional vertex list \"\"\"\n",
    "        self.index = {v:i for i,v in enumerate(vlist)}\n",
    "        self.vertex = {i:v for i,v in enumerate(vlist)}\n",
    "        self.edge = []\n",
    "        self.edgelabel = []\n",
    "        \n",
    "    def addVertex(self, label):\n",
    "        \"\"\" Add a labeled vertex to the graph \"\"\"\n",
    "        index = len(self.index)\n",
    "        self.index[label] = index\n",
    "        self.vertex[index] = label\n",
    "        \n",
    "    def addEdge(self, vsrc, vdst, label='', repeats=True):\n",
    "        \"\"\" Add a directed edge to the graph, with an optional label. \n",
    "        Repeated edges are distinct, unless repeats is set to False. \"\"\"\n",
    "        e = (self.index[vsrc], self.index[vdst])\n",
    "        if (repeats) or (e not in self.edge):\n",
    "            self.edge.append(e)\n",
    "            self.edgelabel.append(label)\n",
    "    \n",
    "    def hamiltonianPath(self):\n",
    "        \"\"\" A Brute-force method for finding a Hamiltonian Path. \n",
    "        Basically, all possible N! paths are enumerated and checked\n",
    "        for edges. Since edges can be reused there are no distictions\n",
    "        made for *which* version of a repeated edge. \"\"\"\n",
    "        for path in itertools.permutations(sorted(self.index.values())):\n",
    "            for i in range(len(path)-1):\n",
    "                if ((path[i],path[i+1]) not in self.edge):\n",
    "                    break\n",
    "            else:\n",
    "                return [self.vertex[i] for i in path]\n",
    "        return []\n",
    "    \n",
    "    def SearchTree(self, path, verticesLeft):\n",
    "        \"\"\" A recursive Branch-and-Bound Hamiltonian Path search. \n",
    "        Paths are extended one node at a time using only available\n",
    "        edges from the graph. \"\"\"\n",
    "        if (len(verticesLeft) == 0):\n",
    "            self.PathV2result = [self.vertex[i] for i in path]\n",
    "            return True\n",
    "        for v in verticesLeft:\n",
    "            if (len(path) == 0) or ((path[-1],v) in self.edge):\n",
    "                if self.SearchTree(path+[v], [r for r in verticesLeft if r != v]):\n",
    "                    return True\n",
    "        return False\n",
    "    \n",
    "    def hamiltonianPathV2(self):\n",
    "        \"\"\" A wrapper function for invoking the Branch-and-Bound \n",
    "        Hamiltonian Path search. \"\"\"\n",
    "        self.PathV2result = []\n",
    "        self.SearchTree([],sorted(self.index.values()))                \n",
    "        return self.PathV2result\n",
    "    \n",
    "    def degrees(self):\n",
    "        \"\"\" Returns two dictionaries with the inDegree and outDegree\n",
    "        of each node from the graph. \"\"\"\n",
    "        inDegree = {}\n",
    "        outDegree = {}\n",
    "        for src, dst in self.edge:\n",
    "            outDegree[src] = outDegree.get(src, 0) + 1\n",
    "            inDegree[dst] = inDegree.get(dst, 0) + 1\n",
    "        return inDegree, outDegree\n",
    "    \n",
    "    def verifyAndGetStart(self):\n",
    "        inDegree, outDegree = self.degrees()\n",
    "        start = 0\n",
    "        end = 0\n",
    "        for vert in self.vertex.keys():\n",
    "            ins = inDegree.get(vert,0)\n",
    "            outs = outDegree.get(vert,0)\n",
    "            if (ins == outs):\n",
    "                continue\n",
    "            elif (ins - outs == 1):\n",
    "                end = vert\n",
    "            elif (outs - ins == 1):\n",
    "                start = vert\n",
    "            else:\n",
    "                start, end = -1, -1\n",
    "                break\n",
    "        if (start >= 0) and (end >= 0):\n",
    "            return start\n",
    "        else:\n",
    "            return -1\n",
    "        \n",
    "    def eulerianPath(self):\n",
    "        graph = [(src,dst) for src,dst in self.edge]\n",
    "        currentVertex = self.verifyAndGetStart()\n",
    "        path = [currentVertex]\n",
    "        # \"next\" is where vertices get inserted into our tour\n",
    "        # it starts at the end (i.e. it is the same as appending),\n",
    "        # but later \"side-trips\" will insert in the middle\n",
    "        next = 1\n",
    "        while len(graph) > 0:\n",
    "            for edge in graph:\n",
    "                if (edge[0] == currentVertex):\n",
    "                    currentVertex = edge[1]\n",
    "                    graph.remove(edge)\n",
    "                    path.insert(next, currentVertex)\n",
    "                    next += 1\n",
    "                    break\n",
    "            else:\n",
    "                for edge in graph:\n",
    "                    try:\n",
    "                        next = path.index(edge[0]) + 1\n",
    "                        currentVertex = edge[0]\n",
    "                        break\n",
    "                    except ValueError:\n",
    "                        continue\n",
    "                else:\n",
    "                    print(\"There is no path!\")\n",
    "                    return False\n",
    "        vpath = [self.vertex[i] for i in path]\n",
    "        return vpath\n",
    "    \n",
    "    def eulerEdges(self, path):\n",
    "        edgeId = {}\n",
    "        for i in range(len(self.edge)):\n",
    "            edgeId[self.edge[i]] = edgeId.get(self.edge[i], []) + [i]\n",
    "        edgeList = []\n",
    "        for i in range(len(path)-1):\n",
    "            edgeList.append(self.edgelabel[edgeId[path[i],path[i+1]].pop()])            \n",
    "        return edgeList\n",
    "    \n",
    "    def render(self, highlightPath=[]):\n",
    "        \"\"\" Outputs a version of the graph that can be rendered\n",
    "        using graphviz tools (http://www.graphviz.org/).\"\"\"\n",
    "        edgeId = {}\n",
    "        for i in range(len(self.edge)):\n",
    "            edgeId[self.edge[i]] = edgeId.get(self.edge[i], []) + [i]\n",
    "        edgeSet = set()\n",
    "        for i in range(len(highlightPath)-1):\n",
    "            src = self.index[highlightPath[i]]\n",
    "            dst = self.index[highlightPath[i+1]]\n",
    "            edgeSet.add(edgeId[src,dst].pop())\n",
    "        result = ''\n",
    "        result += 'digraph {\\n'\n",
    "        result += '   graph [nodesep=2, size=\"10,10\"];\\n'\n",
    "        for index, label in self.vertex.items():\n",
    "            result += '    N%d [shape=\"box\", style=\"rounded\", label=\"%s\"];\\n' % (index, label)\n",
    "        for i, e in enumerate(self.edge):\n",
    "            src, dst = e\n",
    "            result += '    N%d -> N%d' % (src, dst)\n",
    "            label = self.edgelabel[i]\n",
    "            if (len(label) > 0):\n",
    "                if (i in edgeSet):\n",
    "                    result += ' [label=\"%s\", penwidth=3.0]' % (label)\n",
    "                else:\n",
    "                    result += ' [label=\"%s\"]' % (label)\n",
    "            elif (i in edgeSet):\n",
    "                result += ' [penwidth=3.0]'                \n",
    "            result += ';\\n'                \n",
    "        result += '    overlap=false;\\n'\n",
    "        result += '}\\n'\n",
    "        return result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5> Problem 2.1 </h5>\n",
    "A file of 15-mers simulating short reads file <a href=\"https://raw.githubusercontent.com/xiaoyu12/CS614_Jupyter_notebooks/master/data/reads.fa\" download=\"reads.fa\">reads.fas</a> from a genome. Construct a De Bruijin graph using these 15-mers as edges, and assemble the superstring using the eulerianPath() function. Answer the following questions based on the De Bruijin graph and assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell is provided for your own use in answering problem 2. It will not be graded.\n",
    "# Read an example genome from an URL\n",
    "import urllib.request\n",
    "\n",
    "url = \"https://raw.githubusercontent.com/xiaoyu12/CS614_Jupyter_notebooks/master/data/reads.fa\"\n",
    "with urllib.request.urlopen(url) as fp:\n",
    "    headers, reads = loadFasta(fp)\n",
    "    \n",
    "# Use the Graph class defined above and add your own code to answer the questions below. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5>Enter your answers here. THIS CELL WILL BE GRADED. </h5>\n",
    "1. How many distinct nodes appear in the De Bruijn graph that represents these 15-mers as edges? \n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>\n",
    "2. How many nodes are semi-balenced? How many nodes are balanced? \n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>   \n",
    "3. How many are balanced with both in-degrees and out-degrees equal to 1?\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>\n",
    "4. What is the length of assembled superstring following the Eulerian path that can be constructed in the De Bruijn graph?\n",
    "    \n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>\n",
    "5. The assembled superstring should be from the genome <a href=\"https://raw.githubusercontent.com/xiaoyu12/CS614_Jupyter_notebooks/master/data/sequence.fa\" download=\"sequence.fa\">sequence.fa</a>. What is the starting position of this assembled sequence in that genome? Is the assembled sequence on the forward or the reverse-complement strand of the genome?\n",
    "\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Problem 3 (15 points)</h3>\n",
    "In this assignment, you will implement the Gibbs Sampler for motif finding in the GibbsMotifSearch() function. \n",
    "\n",
    "<p>\n",
    "    Make sure run the following cell, which contains useful functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def Profile(s, DNA, k):\n",
    "    \"\"\" Compute the profile of k-mer motifs in a DNA string \n",
    "        s = list of starting indices, 1-based, 0 means ignore, DNA = list of nucleotide strings,\n",
    "        k = Target Motif length\"\"\"\n",
    "    profile = np.zeros((4, k))\n",
    "    for i in range(k):\n",
    "        # loop over string positions\n",
    "        cnt = dict(zip(\"acgt\",(0,0,0,0)))\n",
    "        for j, sval in enumerate(s):\n",
    "            # loop over DNA strands\n",
    "            base = DNA[j][sval+i] \n",
    "            cnt[base] += 1\n",
    "        profile[:, i] = np.fromiter(cnt.values(), dtype=float) / (float)(len(s))\n",
    "    return profile\n",
    "\n",
    "def Score(s, DNA, k):\n",
    "    \"\"\" compute the consensus SCORE of a given k-mer alignment given \n",
    "        offsets into each DNA string. s = list of starting indices.\n",
    "        DNA = list of nucleotide strings. k = Target Motif length\n",
    "    \"\"\"\n",
    "    score = 0\n",
    "    for i in range(k):\n",
    "        # loop over string positions\n",
    "        cnt = dict(zip(\"acgt\",(0,0,0,0)))\n",
    "        for j, sval in enumerate(s):\n",
    "            base = DNA[j][sval+i] \n",
    "            cnt[base] += 1\n",
    "        score += max(cnt.values())\n",
    "    return score\n",
    "\n",
    "\n",
    "def LikelyhoodKmerProfile(kmer, profile):\n",
    "    \"\"\"Given a profile, compute the probability of a kmer string\"\"\"\n",
    "    # check that kmer length matches with profile length\n",
    "    assert len(kmer) == profile.shape[1]\n",
    "    prob = 1.0\n",
    "    idx = dict(zip(\"acgt\", (0, 1, 2, 3)))\n",
    "    for i in range(len(kmer)):\n",
    "        prob = prob * profile[idx[kmer[i]], i]\n",
    "    return prob\n",
    "\n",
    "def ScanAndScoreProfile(DNA, profile):\n",
    "    \"\"\" Given a list of sequences in DNA and a profile,\n",
    "    find the best alignments of the profile in all sequences\"\"\"\n",
    "    bestAlignment = []\n",
    "    k = profile.shape[1] \n",
    "    for seq in DNA:\n",
    "        maxProbProfile = 0\n",
    "        for s in range(len(seq)-k+1):\n",
    "            prob = ScoreKmerProfile(seq[s:s+k], profile)\n",
    "            if (prob > maxProbProfile):\n",
    "                bestS = s\n",
    "                maxProbProfile = prob\n",
    "        bestAlignment.append(bestS)\n",
    "    return bestAlignment\n",
    "\n",
    "def Consensus(s, DNA, k):\n",
    "    \"\"\" compute the consensus k-Motif of an alignment given offsets into each DNA string.\n",
    "            s = list of starting indices, 1-based, 0 means ignore, DNA = list of nucleotide strings,\n",
    "            k = Target Motif length \"\"\"\n",
    "    consensus = ''\n",
    "    for i in range(k):\n",
    "        # loop over string positions\n",
    "        cnt = dict(zip(\"acgt\",(0,0,0,0)))\n",
    "        for j, sval in enumerate(s):\n",
    "            # loop over DNA strands\n",
    "            base = DNA[j][sval+i] \n",
    "            cnt[base] += 1\n",
    "        consensus += max(cnt.items(), key=lambda tup: tup[1])[0]\n",
    "    return consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Problem 3 function. This cell is graded.\n",
    "# You can use functions defined in the previous cell\n",
    "\n",
    "import numpy as np\n",
    "from numpy.random import randint\n",
    "from numpy.random import choice\n",
    "\n",
    "def GibbsMotifSearch(DNA, k, r = 30, N = 5000):\n",
    "    \"\"\" Searches for a k-length motif that appears \n",
    "    in all given DNA sequences. It begins with a\n",
    "    random set of candidate consensus motifs \n",
    "    derived from the data. It refines the motif\n",
    "    using Gibbs sampler until a true consensus emerges.\n",
    "    DNA - list of DNA sequences\n",
    "    k - length of kmer motif\n",
    "    r - number of random initial alignments\n",
    "    N - number of maximum iterations for each Gibbs sampling\n",
    "    \"\"\"\n",
    "    \n",
    "    # initialization\n",
    "    bestAlignment = []\n",
    "    bestScore = 0 \n",
    "    \n",
    "    for sample in range(r):    # Try r different random initial alignments\n",
    "        # Choose a random initial alignment\n",
    "        #   ~~ Your code goes here~~\n",
    "\n",
    "        n = len(DNA)      # number of sequences\n",
    "        for i in range(N):     # Repeat the Gibbs sampling N steps\n",
    "            # select a random sequence for updating. Hint: use the numpy randint() function\n",
    "            #   ~~ Your code goes here~~\n",
    "            \n",
    "            # compute the profile of motifs except for the selected sequence\n",
    "            #   ~~ Your code goes here~~\n",
    "            \n",
    "            # Use the profile to compute the likelyhoods of all k-mers in the select sequence\n",
    "            #   ~~ Your code goes here~~\n",
    "            \n",
    "            # normalize the likelyhoods to compute the sampling probablitiy \n",
    "            # Note: try to check for divide by zero error. \n",
    "            # If total likelyhood is 0, assign equal probability to k-mer positions\n",
    "            #   ~~ Your code goes here~~\n",
    "            \n",
    "            # sample a position in the selected sequence using the probability\n",
    "            # Hint: use the numpy choice() function\n",
    "            #   ~~ Your code goes here~~\n",
    "            \n",
    "            # update the alignment using the sampled position\n",
    "            #   ~~ Your code goes here~~\n",
    "\n",
    "            # compute the score of the new alignment \n",
    "            # if the alignment has a better score than the current best alignment, update the bestAlignment \n",
    "            #   ~~ Your code goes here~~\n",
    "            \n",
    "    \n",
    "    bestScore = Score(bestAlignment, DNA, k)\n",
    "    \n",
    "    return bestAlignment, bestScore\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test your function in this cell\n",
    "seqApprox = [\n",
    "    'tagtggtcttttgagtgtagatctgaagggaaagtatttccaccagttcggggtcacccagcagggcagggtgacttaat',\n",
    "    'cgcgactcggcgctcacagttatcgcacgtttagaccaaaacggagttggatccgaaactggagtttaatcggagtcctt',\n",
    "    'gttacttgtgagcctggttagacccgaaatataattgttggctgcatagcggagctgacatacgagtaggggaaatgcgt',\n",
    "    'aacatcaggctttgattaaacaatttaagcacgtaaatccgaattgacctgatgacaatacggaacatgccggctccggg',\n",
    "    'accaccggataggctgcttattaggtccaaaaggtagtatcgtaataatggctcagccatgtcaatgtgcggcattccac',\n",
    "    'tagattcgaatcgatcgtgtttctccctctgtgggttaacgaggggtccgaccttgctcgcatgtgccgaacttgtaccc',\n",
    "    'gaaatggttcggtgcgatatcaggccgttctcttaacttggcggtgcagatccgaacgtctctggaggggtcgtgcgcta',\n",
    "    'atgtatactagacattctaacgctcgcttattggcggagaccatttgctccactacaagaggctactgtgtagatccgta',\n",
    "    'ttcttacacccttctttagatccaaacctgttggcgccatcttcttttcgagtccttgtacctccatttgctctgatgac',\n",
    "    'ctacctatgtaaaacaacatctactaacgtagtccggtctttcctgatctgccctaacctacaggtcgatccgaaattcg']\n",
    "\n",
    "bestAlignment, bestScore = GibbsMotifSearch(seqApprox, 10, 30, 5000)\n",
    "\n",
    "# Compare it with the result from  MedianMotifSearch()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare your result with that from the MedianMotifSearch(). Are they the same?\n",
    "<p> <span style=\"color: red\"> ~~ Your answer goes here ~~ </span> </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
